libraries/stdlib/js/src/kotlin/text/regex.kt

```diff
@@ -83,19 +83,13 @@ public actual class Regex public actual constructor(pattern: String, options: Se
 
     private var nativeMatchesEntirePattern: RegExp? = null
     private fun initMatchesEntirePattern(): RegExp =
-        nativeMatchesEntirePattern ?: run {
-            if (pattern.startsWith('^') && pattern.endsWith('$'))
-                nativePattern
-            else
-                return RegExp("^${pattern.trimStart('^').trimEnd('$')}$", options.toFlags("gu"))
-        }.also { nativeMatchesEntirePattern = it }
-
+        nativeMatchesEntirePattern ?: RegExp("^(?:$pattern)$", options.toFlags("gu")).also { nativeMatchesEntirePattern = it }
 
     /** Indicates whether the regular expression matches the entire [input]. */
     public actual infix fun matches(input: CharSequence): Boolean {
-        nativePattern.reset()
-        val match = nativePattern.exec(input.toString())
-        return match != null && match.index == 0 && nativePattern.lastIndex == input.length
+        val pattern = initMatchesEntirePattern()
```

Comment:
Compared to the old version, this one will work incorrectly with multiline option turned on (like `Regex("aaaa", setOf(RegexOption.MULTILINE)).matches("aaaa\nbbbb")`): `^(?:aaaa)$` will match only the prefix of the input, and previously we would detect it using the `lastIndex` value.

Reply:
Ouch, good catch, thanks.
I wish JS regexes had `\A` and `\Z` anchors.
I added some more tests and fixed this to work properly with multiline and also with lazy quantifiers.